translator/util/ec2util/ec2util.go

```diff
@@ -67,46 +71,51 @@ func initEC2UtilSingleton() (newInstance *ec2Util) {
 		log.Println("W! [EC2] Sleep until network is up")
 		time.Sleep(1 * time.Second)
 	}
+
 	if !networkUp {
 		log.Println("E! [EC2] No available network interface")
 	}
 
-	ses, err := session.NewSession()
+	err := newInstance.deriveEC2MetadataFromIMDS()
+
 	if err != nil {
-		log.Println("E! [EC2] getting new session info: ", err)
-		return
+		log.Println("E! [EC2] Cannot get EC2 Metadata from IMDS:", err)
 	}
-	md := ec2metadata.New(ses)
 
-	if !md.Available() {
-		log.Println("E! ec2metadata is not available")
-		return
-	}
+	return
+}
 
-	if info, err := md.GetMetadata("instance-id"); err == nil {
-		newInstance.InstanceID = info
-	} else {
-		log.Println("E! getting instance-id from EC2 metadata fail: ", err)
+func (e *ec2Util) deriveEC2MetadataFromIMDS() error {
```

Comment:
So the only two scenarios where we would return an actual error are if:
1. There is some error with creating a session
2.  IMDS is unavailable

We fail silently (with logs) in a case where there is an error trying to fetch specific data from the metadata endpoint. Why is that? There should be more consistency with the behavior of this function. Does it make sense to return a value here? If so, what are the different intended values, and what causes them to be returned?

Reply:
Expanding on this - does it make sense to not return an error if there is a non-nil error when trying to get the hostname from the metadata endpoint? Isn't that value required to be populated?

Reply:
As you mentioned, we would return an actual error for both case. However, does that mean the error will impact the whole process? My answer is no (the same as existing behavior), the return error value only to consolidate the message for the customers. Based on my understanding, the existing behavior for CloudWatchAgent aware that it could be failing in getting the metadata (from host name or from get instance document). However, it does not return error because it can be replaced by the [placeHolder function](https://github.com/aws/amazon-cloudwatch-agent/blob/master/translator/translate/util/placeholderUtil.go#L65-L97). 
So long short story, CloudWatchAgent only expect the failure from those aforementioned case and will always return the values (even if it different intended values because of the replacement with placeholder function) so there are no reason why we would want to return the error when getting one of these metadata endpoint. 

**Note**: This does not apply to EC2Tagger